src/Symfony/Component/String/AbstractString.php

```diff
@@ -0,0 +1,661 @@
+    {
+        if (null === $i = $this->indexOf($needle, $offset)) {
+            $str = clone $this;
+            $str->string = '';
+
+            return $str;
+        }
+
+        if (!$includeNeedle) {
+            $str = clone $this;
+            $str->string = $needle;
+            $i += $str->length();
+        }
+
+        return $this->slice($i);
+    }
+
+    /**
+     * @param string|string[] $needle
+     *
+     * @return static
+     */
+    public function afterLast($needle, bool $includeNeedle = false, int $offset = 0): self
+    {
+        if (null === $i = $this->indexOfLast($needle, $offset)) {
+            $str = clone $this;
+            $str->string = '';
+
+            return $str;
+        }
+
+        if (!$includeNeedle) {
+            $str = clone $this;
+            $str->string = $needle;
+            $i += $str->length();
+        }
+
+        return $this->slice($i);
+    }
+
+    /**
+     * @return static
+     */
+    abstract public function append(string ...$suffix): self;
+
+    /**
+     * @param string|string[] $needle
+     *
+     * @return static
+     */
+    public function before($needle, bool $includeNeedle = false, int $offset = 0): self
+    {
+        if (null === $i = $this->indexOf($needle, $offset)) {
+            $str = clone $this;
+            $str->string = '';
+
+            return $str;
+        }
+
+        if ($includeNeedle) {
+            $str = clone $this;
+            $str->string = $needle;
+            $i += $str->length();
+        }
+
+        return $this->slice(0, $i);
+    }
+
+    /**
+     * @param string|string[] $needle
+     *
+     * @return static
+     */
+    public function beforeLast($needle, bool $includeNeedle = false, int $offset = 0): self
+    {
+        if (null === $i = $this->indexOfLast($needle, $offset)) {
+            $str = clone $this;
+            $str->string = '';
+
+            return $str;
+        }
+
+        if ($includeNeedle) {
+            $str = clone $this;
+            $str->string = $needle;
+            $i += $str->length();
+        }
+
+        return $this->slice(0, $i);
+    }
+
+    /**
+     * @return static
+     */
+    public function camel(): self
+    {
+        $str = clone $this;
+        $str->string = lcfirst(str_replace(' ', '', ucwords(preg_replace('/[^a-zA-Z0-9\x7f-\xff]++/', ' ', $this->string))));
+
+        return $str;
+    }
+
+    /**
+     * @return static[]
+     */
+    abstract public function chunk(int $length = 1): array;
+
+    /**
+     * @return static
+     */
+    public function collapseWhitespace(): self
+    {
+        $str = clone $this;
+        $str->string = trim(preg_replace('/(?:\s{2,}+|[^\S ])/', ' ', $str->string));
+
+        return $str;
+    }
+
+    /**
+     * @param string|string[] $suffix
+     */
+    public function endsWith($suffix): bool
+    {
+        if (!\is_array($suffix) && !$suffix instanceof \Traversable) {
```

Comment:
why not `is_iterable()`?

Reply:
Because it's slower and this code can be found in tight loops so it matters here.